src/Lecture3.hs

```diff
@@ -102,86 +101,104 @@ have a lawful 'Monoid' instance.
+      }
+
+-- | 'List1' is a list that contains at least one element.
 data List1 a = List1 a [a]
-    deriving (Show, Eq)
+  deriving (Show, Eq)
 
 -- | This should be list append.
 instance Semigroup (List1 a) where
+  (<>) :: List1 a -> List1 a -> List1 a
+  (<>) (List1 x xs) (List1 y ys) = List1 x (xs <> (y : ys))
 
+-- | Does 'List1' have the 'Monoid' instance? If no then why?
+-- Answer: List1 does not have a Monoid instance.
+-- The empty element for appending 2 lists is the empty list. List1 being the set of non-empty
+-- lists, [] is not a valid List1. Appending 2 List1 always appends at least 1 element.
+-- instance Monoid (List1 a) where..
 
-{- | Does 'List1' have the 'Monoid' instance? If no then why?
-
-instance Monoid (List1 a) where
--}
-
-{- | When fighting a monster, you can either receive some treasure or
-don't.
--}
+-- | When fighting a monster, you can either receive some treasure or
+-- don't.
 data Treasure a
-    = NoTreasure
-    | SomeTreasure a
-    deriving (Show, Eq)
-
-{- | When you append multiple treasures for fighting multiple
-monsters, you should get a combined treasure and not just the first
-(or last one).
-
-🕯 HINT: You may need to add additional constraints to this instance
-  declaration.
--}
-instance Semigroup (Treasure a) where
-
-
-instance Monoid (Treasure a) where
-
-
-{- | Abstractions are less helpful if we can't write functions that
-use them!
-
-Implement a polymorphic function that takes three elements and appends
-together only different elements.
-
->>> appendDiff3 [1] [3, 2] [0, 5]
-[1,3,2,0,5]
->>> appendDiff3 [4] [2, 2] [2, 2]
-[4,2,2]
->>> appendDiff3 [1 .. 5] [1 .. 5] [1 .. 5]
-[1,2,3,4,5]
->>> appendDiff3 (Product 2) (Product 3) (Product 3)
-Product {getProduct = 6}
-
--}
-appendDiff3 = error "TODO"
+  = NoTreasure
+  | SomeTreasure a
+  deriving (Show, Eq)
+
+-- | When you append multiple treasures for fighting multiple
+-- monsters, you should get a combined treasure and not just the first
+-- (or last one).
+--
+-- 🕯 HINT: You may need to add additional constraints to this instance
+--  declaration.
+instance Semigroup a => Semigroup (Treasure a) where
+  (<>) :: Treasure a -> Treasure a -> Treasure a
+  (<>) NoTreasure t = t
+  (<>) t NoTreasure = t
+  (<>) (SomeTreasure t1) (SomeTreasure t2) = SomeTreasure (t1 <> t2)
+
+instance Semigroup a => Monoid (Treasure a) where
+  mempty :: Treasure a
+  mempty = NoTreasure
+
+-- | Abstractions are less helpful if we can't write functions that
+-- use them!
+--
+-- Implement a polymorphic function that takes three elements and appends
+-- together only different elements.
+--
+-- >>> appendDiff3 [1] [3, 2] [0, 5]
+-- [1,3,2,0,5]
+-- >>> appendDiff3 [4] [2, 2] [2, 2]
+-- [4,2,2]
+-- >>> appendDiff3 [1 .. 5] [1 .. 5] [1 .. 5]
+-- [1,2,3,4,5]
+-- >>> appendDiff3 (Product 2) (Product 3) (Product 3)
+-- Product {getProduct = 6}
+appendDiff3 :: (Monoid a, Eq a) => a -> a -> a -> a
+appendDiff3 x y z = fold (nub [x, y, z])
```

Comment:
That's a clever and concise implementation! 👏🏻 

However, the type signature is more strict than required. Your function works with `Monoid`. But not every type can have a `Monoid` instance. This means, that `appendDiff3` can't work with types like `List1`. But if you change the constraint from `Monoid` to `Semigroup`, this function will now work. Though, relaxing the constraint will require changing the implementation as well 🚧

Reply:
I wrote a less constrained implementation, and almost as concise, with ```Semigroup``` instead. Thanks for pointing that out.